packages/headless/src/controllers/instant-results/instant-results.ts

```diff
@@ -0,0 +1,131 @@
+import {SearchEngine} from '../../app/search-engine/search-engine';
+import {Result} from '../../api/search/search/result';
+import {InstantResultSection} from '../../state/state-sections';
+import {buildController, Controller} from '../controller/headless-controller';
+import {instantResults} from '../../app/reducers';
+import {loadReducerError} from '../../utils/errors';
+import {validateOptions} from '../../utils/validate-payload';
+import {
+  fetchInstantResults,
+  registerInstantResults,
+  updateInstantResultsQuery,
+} from '../../features/instant-results/instant-results-actions';
+import {
+  InstantResultOptions,
+  instantResultsOptionsSchema,
+} from './instant-results-options';
+import {SearchAPIErrorWithStatusCode} from '../../api/search/search-api-error-response';
+import {SerializedError} from '@reduxjs/toolkit';
+
+export interface InstantResultProps {
+  options: InstantResultOptions;
+}
+
+export interface InstantResults extends Controller {
+  /**
+   * Updates the instant results query and shows instant results for it.
+   *
+   * @param q - The string value to get instant results for.
+   *
+   */
+  updateQuery(q: string): void;
+  /**
+   * The state of the `InstantResults` controller.
+   */
+  state: InstantResultsState;
+}
+
+export interface InstantResultsState {
+  /**
+   * The current query for instant results.
+   */
+  q: string;
+  /**
+   * The instant results for the current query in instant results.
+   */
+  results: Result[];
+  /**
+   * Determines if a search is in progress for the current instant results query.
+   */
+  isLoading: boolean;
+  /**
+   * Determines if a search is in progress for the current instant results query.
+   */
+  error: SearchAPIErrorWithStatusCode | SerializedError | null;
+}
+
+/**
+ * Creates a `InstantResults` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `InstantResults` properties.
+ * @returns A `InstantResults` controller instance.
+ */
+export function buildInstantResults(
+  engine: SearchEngine,
+  props: InstantResultProps
+): InstantResults {
+  if (!loadInstantResultsReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+  const getState = () => engine.state;
+
+  const options: Required<InstantResultOptions> = {
+    ...props.options,
+  };
+
+  validateOptions(
+    engine,
+    instantResultsOptionsSchema,
+    options,
+    'buildInstantResults'
+  );
+
+  const searchBoxId = options.searchBoxId;
+  dispatch(registerInstantResults({id: searchBoxId}));
+
+  const getStateForSearchBox = () => getState().instantResults[searchBoxId];
+
+  const getCached = (q: string) => getStateForSearchBox().cache[q];
+  const getQ = () => getStateForSearchBox().q;
+  const getResults = () => {
+    const cached = getCached(getQ());
```

Comment:
The cache is a good idea, but it might need to be a bit need to become more sophisticated than this.

* Is there a max size ?
* Is there a max duration ?
* Should an implementer be able to clear the cache if they want to ?
* Should an implementer be able to configure the cache parameters if they want to ?
* What kind of problem/questions can we expect with this cache ? For example, if someone is doing an implementation and they are trying to debug their instant result, and they don't see "new content" from their index because they did not refresh the page, how will they know there's a cache ? Is this overthinking or is it a true problem ?

We can leave it like that for now, but food for thoughts.

Reply:
Enhance later is 👍 .

Reply:
I got excited and added the `expiresAt` key in the latest commits